### PR TITLE
Native desktop packaging

### PR DESCRIPTION
**Pull request by AlexeyTsvetkov**

### CODE REVIEW

File: experimental/examples/falling-balls-mpp/build.gradle.kts [hunk 1]
```diff
@@ -166,6 +166,15 @@ kotlin {
     }
 }
 
+compose.desktop.nativeApplication {
+    targets(kotlin.targets.getByName("macosX64"))
```
Comment (olonho) [inline experimental/examples/falling-balls-mpp/build.gradle.kts:170]: Could we build ARM targets as well?

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/dsl/JvmApplicationDistributions.kt [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+abstract class JvmApplicationDistributions : AbstractDistributions() {
+    var modules = arrayListOf(*DEFAULT_RUNTIME_MODULES)
+    fun modules(vararg modules: String) {
+        this.modules.addAll(modules.toList())
+    }
+    var includeAllModules: Boolean = false
+
+    val linux: LinuxPlatformSettings = objects.newInstance(LinuxPlatformSettings::class.java)
+    open fun linux(fn: Action<LinuxPlatformSettings>) {
```
Comment (olonho) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/dsl/JvmApplicationDistributions.kt:22]: Maybe make `JvmLinuxPlatformSettings` and `JvmWindowsPlatformSettings` instead?

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/tasks/AbstractNativeMacApplicationPackageDmgTask.kt [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        }
+        finalizeImage(tmpImage, finalImage)
+        logger.lifecycle("The distribution is written to ${finalImage.canonicalPath}")
+    }
+
+    private fun createImage(volumeName: String, imageFile: File, srcDir: File) {
+        var size = srcDir.walk().filter { it.isFile }.sumOf { it.length() }
+        size += 10 * 1024 * 1024
+
+        hdiutil(
```
Comment (olonho) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/tasks/AbstractNativeMacApplicationPackageDmgTask.kt:52]: Wow, that's easy.

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/tasks/AbstractNativeMacApplicationPackageDmgTask.kt [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            "-noautoopen",
+            imageFile.absolutePath
+        )
+        Thread.sleep(3000)
+        var device: String? = null
+        val volume: String = output.substring(output.lastIndexOf("/Volumes/$volumeName")).trim()
+
+        for (entry in output.split("\\s+".toRegex())) {
+            if (entry.startsWith("/dev/")) {
+                device = entry
```
Comment (olonho) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/tasks/AbstractNativeMacApplicationPackageDmgTask.kt:78]: Could there be multiple entries like that?

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/tasks/AbstractUnpackDefaultComposeApplicationResourcesTask.kt [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import org.gradle.api.provider.Provider
+import org.gradle.api.tasks.Internal
+import org.gradle.api.tasks.OutputDirectory
+import org.gradle.api.tasks.TaskAction
+import org.jetbrains.compose.ComposeBuildConfig
+import org.jetbrains.compose.desktop.application.internal.ioFile
+
+abstract class AbstractUnpackDefaultComposeApplicationResourcesTask : AbstractComposeDesktopTask() {
+    internal class DefaultResourcesProvider(resourcesRootDir: Provider<Directory>) {
+        val macIcon: Provider<RegularFile> = resourcesRootDir.map { it.file("default-icon-mac.icns") }
```
Comment (olonho) [inline gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/tasks/AbstractUnpackDefaultComposeApplicationResourcesTask.kt:20]: Are those names somehow documented?

**Review by olonho — APPROVED** [review]
Looks good, modulo few nits.

File: experimental/examples/falling-balls-mpp/build.gradle.kts (hunk 1 above)
Comment (AlexeyTsvetkov) [reply experimental/examples/falling-balls-mpp/build.gradle.kts:186]: Yes

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/tasks/AbstractNativeMacApplicationPackageDmgTask.kt (hunk 4 above)
Comment (AlexeyTsvetkov) [reply gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/application/tasks/AbstractNativeMacApplicationPackageDmgTask.kt:78]: Maybe, fixed

File: gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/tasks/AbstractUnpackDefaultComposeApplicationResourcesTask.kt (hunk 5 above)
Comment (AlexeyTsvetkov) [reply gradle-plugins/compose/src/main/kotlin/org/jetbrains/compose/desktop/tasks/AbstractUnpackDefaultComposeApplicationResourcesTask.kt:20]: No, and I don't think either of them is a part of public contract. After all, an icon is copied to a `<APP_NAME>.app` as `<APP_NAME>.icns`

**Pull request merged**